k8s/docs/change-control-process.md

```diff
@@ -0,0 +1,106 @@
+   system may not be possible due to the change control software preventing or
+   automatically undoing the changes.
+
+#### 1 - Staging, Patch Files, and the main Branch
+
+The spack infrastructure repository has a number of key branches that serve
+different purposes.  First among these is the `main` branch.  This branch
+contains the latest head of development and is the branch that most pull
+requests will be made against.  The overall change control process starts with
+these pull requests.
+
+Among the files in the repository are a number of special config maps
+representing "staged resources" - resources that are deployed for staging
+purposes.  Such resources are automatically deployed from the `main` branch and
+updated when new commits are merged to the `main` branch.
+
+#### 2 - Production and Maintenance Windows
+
+The contents of the `production` branch represent the resources that are
+deployed and managed for production services.  The `production` branch typically
+lags behind the `main` branch and is periodically brought up-to-date with the
+`main` branch.  In general, these pushes to `production` are done during a
+maintenance window - a predetermined window of time within which production
+services may become unavailable.  While there are no formal service level
+agreements for the production services, the regularly-scheduled maintenance
+window affords us the option to perform service-disrupting maintanenace work
+within a specific period of time while providing service availability on a
+best-effort basis outside of that period.  Currently, the maintenance windows
+occur on the first and third Friday of every month, starting on midnight,
+eastern time, on that Friday and ending on midnight, eastern time, the next day.
+
+#### 3 - The flux Branch
+
+The `flux` branch is a branch that is automatically managed by middleware
+running on the cluster.  Its contents represent the exact resources that are
+provisioned on the cluster at any given time and are generated by combing the
+contents of the `main` and `production` branches for staged and production
+resources, respectively.
+
+#### 4 - Change Control Process Outline
+
+ 1. One or more pull requests are opened against the `main` branch and
+    eventually merged.
+
+    - The contents of these pull requests are iterated on through some
+      combination of local and live experimentation.
+
+ 2. The state of the `production` branch is modified during a maintenance
+    window.
+
+    - First, a new pull request that brings `production` up-to-date with `main`
+      is opened and merged.
+
+    - Cluster middleware triggers changes to the `flux` branch.
+
+    - Cluster middleware responds to changes in the `flux` branch by updating
+      cluster resources.
+
+    - The state of the cluster and services are reviewed for proper operation.
+
+ 3. If there are any issues that arise as a result of step 2, then remedial
```

Comment:
I would add in here that it's possible to have a local clone where you are interacting with the cluster, and the PR opened for a long period of time (e.g., I don't think I've ever had any of my PRs actually merged, and this is understandable because it's sort of "developing and tweaking as I go" work.  But it's good to set expectations about merging/not without branding not merging as "this isn't good enough" when that's not the case.

Reply:
That's a great point.

Speaking of which... we might want to start thinking about getting those PRs merged so that they can benefit from this process.  Iterating on your services would take a little more work, but it's something we should do once you feel like you've gotten those PRs to a good, stable state.

Reply:
What would you define as stable? I will likely continue to tweak the image tag if there are changes to the server - that's definitely died down from its original frequency, but could happen again after more discussion during a team meeting. The other changes that might be likely are upping the pvc sizes (like we did early this week, successfully!) I'm happy to leave them in a PR state if these kind of changes make the deployments not yet stable.

And for drift server ping @alecbcs!

Reply:
So, I added some text that touches on this, but it really depends on you.  The question is "are your services at a point where you'd like their deployment state to be part of this change control process?"  The upside is that your deployment state is tied to the repo state, so there's never a question of what version of your application is currently deployed.  Also, updates to your services are rolled out on a regular schedule using a process that allows for easy roll back in case there are any issues with new changes.

The downside is that you no longer have the flexibility of just tweaking your application deployment live.  You would have to set up a separate deployment to have something to tweak.

Reply:
It sounds easier then to leave as is! Let's do that :)